File: Heaps/cpp/minHeapImplementation.cpp
// Problem Statement: Implement min heap

// Problem Link: https://www.codingninjas.com/studio/problems/min-heap-implementation_5480527?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

#include <iostream>
#include <vector>
using namespace std;

class MinHeap {
private:
    vector<int> heap;
    int size;
    int capacity;

    void heapify(int i) {
        int smallest = i;
        int leftChild = 2 * i + 1;
        int rightChild = 2 * i + 2;

        if (leftChild < size && heap[leftChild] < heap[smallest]) {
            smallest = leftChild;
        }

        if (rightChild < size && heap[rightChild] < heap[smallest]) {
            smallest = rightChild;
        }

        if (i != smallest) {
            swap(heap[i], heap[smallest]);
            heapify(smallest);
        }
    }

public:
    MinHeap(int size) {
        this->capacity = size;
        this->heap.resize(size);
        this->size = 0;
    }

    int extractMinElement() {
        if (size == 0) {
            return -1; // Heap is empty
        }

        int minValue = heap[0];
        heap[0] = heap[size - 1];
        size--;
        heapify(0);
        return minValue;
    }

    void deleteElement(int ind) {
        if (ind < 0 || ind >= size) {
            return; // Index out of bounds
        }

        heap[ind] = heap[size - 1];
        size--;
        heapify(ind);
    }

    void insert(int val) {
        if (size == capacity) {
            return; // Heap is full
        }

        heap[size] = val;
        size++;
        int currentIndex = size - 1;

        while (currentIndex > 0) {
            int parentIndex = (currentIndex - 1) / 2;
            if (heap[parentIndex] > heap[currentIndex]) {
                swap(heap[parentIndex], heap[currentIndex]);
                currentIndex = parentIndex;
            } else {
                break;
            }
        }
    }
};

File: Heaps/python/minHeapImplementation.py
# Problem Statement: Implement min heap

# Problem Link: https://www.codingninjas.com/studio/problems/min-heap-implementation_5480527?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

class MinHeap:
    def __init__(self, size):
        self.capacity = size
        self.heap = [0] * size
        self.size = 0

    def extract_min_element(self):
        if self.size == 0:
            return -1
            
        min_value = self.heap[0]
        self.heap[0] = self.heap[self.size - 1]
        self.size -= 1
        self._heapify(0)
        return min_value

    def delete_element(self, ind):
        if ind < 0 or ind >= self.size:
            return
            
        self.heap[ind] = self.heap[self.size - 1]
        self.size -= 1
        self._heapify(ind)

    def insert(self, val):
        if self.size == self.capacity:
            return
            
        self.heap[self.size] = val
        self.size += 1
        current_index = self.size - 1
        
        while current_index > 0:
            parent_index = (current_index - 1) // 2
            if self.heap[parent_index] > self.heap[current_index]:
                self.heap[parent_index], self.heap[current_index] = self.heap[current_index], self.heap[parent_index]
                current_index = parent_index
            else:
                break

    def _heapify(self, i):
        smallest = i
        left_child = 2 * i + 1
        right_child = 2 * i + 2
        
        if left_child < self.size and self.heap[left_child] < self.heap[smallest]:
            smallest = left_child
            
        if right_child < self.size and self.heap[right_child] < self.heap[smallest]:
            smallest = right_child
            
        if i != smallest:
            self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
            self._heapify(smallest)
